Migrate to use systemd events

jordigilh opened the pull request:
- Changes to use systemd events instead of relying on podman events. 
- Refactored systemd functions for clarity.

Note: I tried to create an interface layer between dbus and the event listener so I could add unit tests. Unfortunately the initial effort translated into creating extra complexity because the dbus package is not providing interfaces. Suggestions are welcome.

jordigilh commented:
Failing to properly configure the file permissions for `volumes` directory when run in a container. No issues when run in a VM. Investigating.

jordigilh commented:
> Failing to properly configure the file permissions for `volumes` directory when run in a container. No issues when run in a VM. Investigating.

Fixed in https://github.com/project-flotta/flotta-operator/pull/290 . It is not a showstopper for this PR as it only impacts when running a custom RPM for the end to end tests.

tupyy commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,116 @@
+	dbusErrCh  <-chan error
+}
+
+func NewEventListener(observerCh chan *Event) *EventListener {
```
being the producer, I would own this channel. I can make sure it is properly initialized

jordigilh replied:
Agreed. Let me refactor this.

jordigilh replied:
Done. Please take a look.

tupyy commented on internal/service/systemd.go:
```diff
@@ -55,12 +64,13 @@ type SystemdManager interface {
 }
 
-func NewSystemdManager(configDir string) (SystemdManager, error) {
+func NewSystemdManager(configDir string, observerCh chan *Event) (SystemdManager, error) {
```
who owns this channel? Why is coming from the upper layer when it is used internally?

jordigilh replied:
right now it's created by the workloadManager as part of the new instance logic:
https://github.com/project-flotta/flotta-device-worker/pull/217/files#diff-27f5b4db08ed174b7b18101c679d78c4dcf69db6fd8d6772f6a462f0fd6ae453R86

I'm going to refator this and pass the listener to the `NewSystemdManager` instead of hiding it inside that function. I think you will like it more :wink:

jordigilh replied:
Done. Please take a look.

jordigilh commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,116 @@
+
+func (e *EventListener) Listen() {
+	e.dbusCh, e.dbusErrCh = e.set.Subscribe()
+	go func() {
```
sounds good, I'll change the caller to trigger the routine as suggested.

jordigilh replied:
done, please take a look.

masayag commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,116 @@
+
+func (e *EventListener) Remove(workloadName string) {
+	name := fmt.Sprintf("%s.service", workloadName)
+	log.Debugf("Removing service for events %s", name)
```
should the message say `for workload`?

jordigilh replied:
I was not sure about this since we're removing the systemd service from the filter. I show the message related to workload removed elsewhere:
https://github.com/project-flotta/flotta-device-worker/pull/217/files#diff-27f5b4db08ed174b7b18101c679d78c4dcf69db6fd8d6772f6a462f0fd6ae453R124

masayag commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,116 @@
+
+func (e *EventListener) Add(workloadName string) {
+	name := fmt.Sprintf("%s.service", workloadName)
+	log.Debugf("Adding service for events %s", name)
```
should the message say `for workload`?

jordigilh replied:
Same as earlier, but in this case the message for workload started is generated here:
https://github.com/project-flotta/flotta-device-worker/pull/217/files#diff-27f5b4db08ed174b7b18101c679d78c4dcf69db6fd8d6772f6a462f0fd6ae453R115

masayag commented on internal/workload/wrapper.go:
```diff
@@ -131,7 +141,7 @@ func (ww *Workload) RegisterObserver(observer Observer) {
 func (ww *Workload) Init() error {
 	// Enable auto-update podman timer:
-	svc, err := service.NewSystemdRootless("podman-auto-update", nil, false)
+	svc, err := service.NewSystemd("podman-auto-update", nil, service.SystemBus, nil)
```
question: why podman-auto-update need SystemBus and not UserBus?
(I'm aware this part is just a refactor, but still trying to understand this part since we handle the other podman's service as UserBus)

jordigilh replied:
No idea and both are valid services. I did not want to change it because I was unsure and I didn't want to make more changes than needed. But in essence I agree that it looks to be more for the `userBus` than the `systemBus`, maybe @machacekondra can chime in and share his thoughts as well.

masayag commented on internal/service/systemd.go:
```diff
@@ -20,12 +20,20 @@ const (
 	DefaultRestartTimeout = 15
 	TimerSuffix           = ".timer"
 	ServiceSuffix         = ".service"
+	DefaultNameSeparator  = "-"
```
this isn't used.

jordigilh replied:
Deleted. I wonder why no checks flagged this in the PR workflow.

masayag commented on internal/workload/wrapper.go:
```diff
@@ -273,18 +270,19 @@ func (ww *Workload) removeService(workloadName string) error {
-
+	err := svc.Stop()
+	if err != nil {
+		log.Errorf("unable to stop service %s:%s", workloadName, err)
```
s/unable/Unable ?

masayag commented on internal/workload/wrapper.go:
```diff
@@ -273,18 +270,19 @@ func (ww *Workload) removeService(workloadName string) error {
+	err = ww.serviceManager.Remove(svc)
 	if err != nil {
-		return nil
+		log.Errorf("unable to remove service from serviceManager %s:%s", workloadName, err)
```
s/unable/Unable ?

jordigilh replied:
Done

eloycoto commented:
@jordigilh what I mentioned yesterday regarding restarts:

Initialize all as normal, and deploy a workload; you will get the info here: 
https://github.com/project-flotta/flotta-device-worker/blob/main/internal/metrics/workload.go#L55

After all the pods are running, you need to restart yggdrasild
```
systemctl restart yggdrasild
```
there is no notification that the workloadStarted, so the metrics are not updated at all, and it'll not scrape as it should.

jordigilh commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+func (e *EventListener) Add(workloadName string) {
+	name := fmt.Sprintf("%s.service", workloadName)
```
done

jordigilh commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,116 @@
+func (e *EventListener) Add(workloadName string) {
+	name := fmt.Sprintf("%s.service", workloadName)
+	log.Debugf("Adding service for events %s", name)
+	if !e.set.Contains(name) {
```
it's a safeguard just in case the service was already part of the filter. In theory this should never happen based on the nominal flow. If a service is replaced it will be first removed and then readded by the events.

jordigilh replied:
Ended up removing it with the new logic to watch for file system events. The case where the service could be added twice is not possible anymore.

jordigilh commented:
> @jordigilh what I mentioned yesterday regarding restarts:
> 
> Initialize all as normal, and deploy a workload; you will get the info here: https://github.com/project-flotta/flotta-device-worker/blob/main/internal/metrics/workload.go#L55
> 
> After all the pods are running, you need to restart yggdrasild
> 
> ```
> systemctl restart yggdrasild
> ```
> 
> there is no notification that the workloadStarted, so the metrics are not updated at all, and it'll not scrape as it should.

That's weird, I'm actually seeing it:
```
[root@fedora ~]# systemctl restart yggrasild
Failed to restart yggrasild.service: Unit yggrasild.service not found.
[root@fedora ~]# systemctl restart yggdrasild
[root@fedora ~]# Aug 02 11:14:39 fedora systemd[1]: Stopping yggdrasild.service - yggdrasil daemon...
Aug 02 11:14:39 fedora systemd[1]: yggdrasild.service: Killing process 1098 (device-worker) with signal SIGKILL.
Aug 02 11:14:39 fedora systemd[1]: yggdrasild.service: Deactivated successfully.
Aug 02 11:14:39 fedora systemd[1]: Stopped yggdrasild.service - yggdrasil daemon.
Aug 02 11:14:39 fedora systemd[1]: yggdrasild.service: Consumed 2.286s CPU time.
Aug 02 11:14:39 fedora systemd[1]: Started yggdrasild.service - yggdrasil daemon.
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 starting yggdrasild version 0.2.99
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 cannot start watching '/etc/yggdrasil/tags.toml': lstat /etc/yggdrasil/tags.toml: no such file or directory
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 listening on socket: @yggd-dispatcher-vTDVZH
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 cannot route message  123d34f8-412f-4eba-b718-eddffc010444 to directive: device
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 worker registered: {pid:1225 handler:device addr:@ygg-device-pqOEQv features:map[] detachedContent:false}
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 /usr/libexec/yggdrasil/device-worker: Data directory: /etc/yggdrasil/device
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 /usr/libexec/yggdrasil/device-worker: device config file: /etc/yggdrasil/device/device-config.json
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 /usr/libexec/yggdrasil/device-worker: started targetMetric for workload 'system'
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 /usr/libexec/yggdrasil/device-worker: started targetMetric for workload 'data transfer'
Aug 02 11:14:39 fedora yggdrasild[1219]: [yggdrasild] 2022/08/02 11:14:39 /usr/libexec/yggdrasil/device-worker: Service for workload mylog started
```

jordigilh commented:
> maybe the channel is not initialized at that time?

In that case the message should not have been received by the wrapper. In my case I'm testing with a workload and an edgedevice that has no metrics nor logs, so maybe that's why there are no other messages.
This is the log output when creating the workload in the first place:
```
Aug 02 11:33:41 fedora yggdrasild[3484]: [yggdrasild] 2022/08/02 11:33:41 /usr/libexec/yggdrasil/device-worker: updating configuration. New config: {"configuration":{"heartbeat":{"hardware_profile":{},"period_seconds":60},"metrics":{"receiver":{"request_num_samples":30000,"timeout_seconds":10}},"mounts":[]},"device_id":"4233c45699b644b79107306e74bccbc5","version":"206051","workloads":[{"name":"mylog","namespace":"default","specification":"containers:\n- image: docker.io/eloycoto/logexample\n  name: log\n  resources: {}\n"}]}
Aug 02 11:33:41 fedora yggdrasild[3484]: [yggdrasild] 2022/08/02 11:33:41 /usr/libexec/yggdrasil/device-worker: Old config: {"configuration":{"heartbeat":{"hardware_profile":{},"period_seconds":60},"metrics":{"receiver":{"request_num_samples":30000,"timeout_seconds":10}},"mounts":[]},"device_id":"4233c45699b644b79107306e74bccbc5","version":"205749"}
Aug 02 11:33:41 fedora yggdrasild[3484]: [yggdrasild] 2022/08/02 11:33:41 /usr/libexec/yggdrasil/device-worker: Creating service for mylog
Aug 02 11:33:41 fedora yggdrasild[3484]: [yggdrasild] 2022/08/02 11:33:41 /usr/libexec/yggdrasil/device-worker: writing new systemd file for 'container-mylog-log' on '/var/home/flotta/.config/systemd/user/container-mylog-log.service'
Aug 02 11:33:41 fedora yggdrasild[3484]: [yggdrasild] 2022/08/02 11:33:41 /usr/libexec/yggdrasil/device-worker: writing new systemd file for 'mylog' on '/var/home/flotta/.config/systemd/user/mylog.service'
Aug 02 11:33:42 fedora yggdrasild[3484]: [yggdrasild] 2022/08/02 11:33:42 /usr/libexec/yggdrasil/device-worker: Adding service for listener for mylog
Aug 02 11:33:42 fedora yggdrasild[3484]: [yggdrasild] 2022/08/02 11:33:42 /usr/libexec/yggdrasil/device-worker: Service for workload mylog started
```
Notice that there is no additional messages for metrics or logs after `Service for workload mylog started`

eloycoto commented:
This is how I'm testing, and that should be received, but it's not. 
https://drive.google.com/file/d/13sfK0L7eACQRhZInzAGfQWtPcKvMHZZ0/view?usp=sharing

jordigilh commented:
> This is how I'm testing, and that should be received, but it's not. 
> 
> https://drive.google.com/file/d/13sfK0L7eACQRhZInzAGfQWtPcKvMHZZ0/view?usp=sharing
> 
> 
> 
> 
> 
> 

Confirmed. The problem is that the observers are not instantiated before the events are triggered.  I will refactor the code to init the observers before the event listener is instantiated

jordigilh commented:
@eloycoto can you try again. I made it Eloy's `fail-proof` now :)

jordigilh commented:
@masayag @tupyy significant refactoring made, so you might want to take another look at the whole thing.

jordigilh commented on cmd/device-worker/main.go:
```diff
@@ -141,6 +142,16 @@ func main() {
+	}
+	// Start listening to systemd bus events. These events generated from here can start being processed once all observers have been initialized, otherwise
+	// we risk in missing observers whilst processing the events.
+	go listener.Listen()
```
I don't think we miss any because the way the events work in systemd is that the first time you tap into the bus you get an event about the service's current status. Subsequent events for that service will be based on a diff from the previous status.

You can try this use case by restarting yggdrasild and you will notice that the workloads are detected as running and the event is propagated to the workload manager once it starts listening.

masayag replied:
thanks for the clarification. So not a concern at all.

jordigilh commented:
@masayag @eloycoto I've refactored the code and removed the FS Notify logic. Instead the event listener is now an Observer and taps into the device config changes to be aware of which services to watch events for.  Take a look at the last commit if you don't want to be bothered with the rest.

One thing that still bothers me is that we don't have any means to track other enabled user services, in case somehow a new service is added to the user outside of the agent's control.

tupyy commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,152 @@
+	removed unitSubState = "removed"
+)
+
+var ()
```
pls remove it

jordigilh replied:
Done.

eloycoto commented on cmd/device-worker/main.go:
```diff
@@ -20,7 +20,8 @@ import (
 	"github.com/project-flotta/flotta-device-worker/internal/server"
-	workload2 "github.com/project-flotta/flotta-device-worker/internal/workload"
+	"github.com/project-flotta/flotta-device-worker/internal/service"
+	workload "github.com/project-flotta/flotta-device-worker/internal/workload"
```
this workload alias is not needed at all, no?

jordigilh replied:
yeah, will remove it.

eloycoto commented on internal/workload/wrapper.go:
```diff
@@ -376,3 +343,35 @@ func getHostPorts(workload *v1.Pod) ([]int32, error) {
+			log.Infof("Service for workload %s started", event.WorkloadName)
+			report, err := w.podManager.GetPodReportForPodName(event.WorkloadName)
+			if err != nil {
+				log.Errorf("unable to get pod report for workload %s:%v", event.WorkloadName, err)
```
I will workload '%s' as line 364

jordigilh replied:
done

eloycoto commented on internal/workload/wrapper.go:
```diff
@@ -376,3 +343,35 @@ func getHostPorts(workload *v1.Pod) ([]int32, error) {
+				observer.WorkloadStarted(event.WorkloadName, []*podman.PodReport{report})
+			}
+		case service.EventStopped:
+			log.Infof("Service for workload %s stopped", event.WorkloadName)
```
'%s'

jordigilh replied:
done

jordigilh commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,172 @@
+					continue
+				}
+				state := translateUnitSubStatus(unit)
+				log.Debugf("Systemd service for workload %s transitioned to sub state %s\n", n, state)
```
It can be too verbose with a few workloads in place. I had them all in Info and it was dumping too much info.

jordigilh commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,172 @@
+	log.Infof("Starting DBus event listener")
+	conn, err := newDbusConnection(UserBus)
+	if err != nil {
+		return err
```
I was not sure what would be the added value. Any suggestion to what to append?

jordigilh replied:
ok I added this:
```
fmt.Errorf("error while starting event listener: %v", err)
```

jordigilh commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,172 @@
+		svcName := DefaultServiceName(wl.Name)
+		if !e.contains(svcName) {
+			e.add(svcName)
+		}
```
Ok will add.

jordigilh replied:
done

jordigilh commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,172 @@
+				log.Debugf("Captured DBus event for %s: %v+", name, unit)
+				n, err := extractWorkloadName(name)
+				if err != nil {
+					log.Error(err)
```
any suggestions? The message already contains the service name

jordigilh commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,172 @@
+					log.Debugf("Sending stop event to observer channel for workload %s", n)
+					e.eventCh <- &Event{WorkloadName: n, Type: EventStopped}
+				default:
+					log.Debugf("Ignoring unit sub state for service %s: %s", name, unit.SubState)
```
There are a few sub states we ignore because they are not relevant to our needs:
- post-start
- post-stop
- pre-start
...
The ones we capture are the ones we want to take action.

jordigilh commented on internal/service/event_listener.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+func (e *DBusEventListener) contains(serviceName string) bool {
+	e.lock.Lock()
```
Sounds good.

jordigilh commented on internal/workload/podman/podman.go:
```diff
@@ -15,9 +15,9 @@ import (
 	"github.com/blang/semver"
 	"github.com/go-openapi/swag"
 	"github.com/project-flotta/flotta-device-worker/internal/service"
+	api "github.com/project-flotta/flotta-device-worker/internal/workload/api"
```
done

jordigilh commented:
@eloycoto I've addressed some of your comments, there are a few that I've replied. PTAL when you have spare time.

jordigilh commented on internal/service/systemd.go:
```diff
@@ -267,7 +262,8 @@ func (s *systemd) Stop() error {
 
 func (s *systemd) Enable() error {
-	conn, err := newDbusConnection(s.Rootless)
+	log.Debugf("Enabling service %s", s.Name)
```
done with %s
